protocols/direct-funding.go

```diff
@@ -0,0 +1,378 @@
+	// todo: []string{"Objective" + s.ChannelId.String() + "complete"} as SideEffects{}
+	return updated, NoSideEffects, WaitingForNothing, nil
+}
+
+//  Private methods on the DirectFundingObjectiveState
+
+// prefundComplete returns true if all participants have signed a prefund state, as reflected by the extended state
+func (s DirectFundingObjectiveState) prefundComplete() bool {
+	for _, index := range s.ParticipantIndex {
+		if !s.PreFundSigned[index] {
+			return false
+		}
+	}
+	return true
+}
+
+// postfundComplete returns true if all participants have signed a postfund state, as reflected by the extended state
+func (s DirectFundingObjectiveState) postfundComplete() bool {
+	for _, index := range s.ParticipantIndex {
+		if !s.PostFundSigned[index] {
+			return false
+		}
+	}
+	return true
+}
+
+// fundingComplete returns true if the supplied onChainHoldings are greater than or equal to the threshold for being fully funded.
+func (s DirectFundingObjectiveState) fundingComplete(onChainHoldings types.Funds) bool {
+	for asset, threshold := range s.FullyFundedThreshold {
+		chainHolding, ok := onChainHoldings[asset]
+
+		if !ok {
+			return false
+		}
+
+		if gt(threshold, chainHolding) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// safeToDeposit returns true if the supplied onChainHoldings are greater than or equal to the threshold for safety.
+func (s DirectFundingObjectiveState) safeToDeposit(onChainHoldings types.Funds) bool {
+	for asset, safetyThreshold := range s.MyDepositSafetyThreshold {
+		chainHolding, ok := onChainHoldings[asset]
+
+		if !ok {
+			return false
+		}
+
+		if gt(safetyThreshold, chainHolding) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// amountToDeposit computes the appropriate amount to deposit using the supplied onChainHolding
+func (s DirectFundingObjectiveState) amountToDeposit(onChainHoldings types.Funds) types.Funds {
+	deposits := make(types.Funds, len(onChainHoldings))
+
+	for asset, holding := range onChainHoldings {
+		deposits[asset] = big.NewInt(0).Sub(s.MyDepositTarget[asset], holding)
+	}
+
+	return deposits
+}
+
+// SignatureReceived updates the objective's cache of which participants have signed which states
+func (s DirectFundingObjectiveState) signatureRecieved(signature state.Signature, turnNum int) (DirectFundingObjectiveState, error) {
+	updated := s.clone()
+
+	signer, err := updated.ExpectedStates[turnNum].RecoverSigner(signature)
+	index, ok := updated.ParticipantIndex[signer]
+
+	if ok && err == nil {
+		if turnNum == 0 {
+			updated.PreFundSigned[index] = true
+		} else if turnNum == 1 {
+			updated.PostFundSigned[index] = true
+		}
+	}
+
+	return updated, nil
+}
+
+// todo: is this sufficient? Particularly: s has pointer members (*big.Int)
+func (s DirectFundingObjectiveState) clone() DirectFundingObjectiveState {
+	return s
+}
+
+// utilities
+
+func gte(a *big.Int, b *big.Int) bool {
+	return a.Cmp(b) > -1
+}
+
+func gt(a *big.Int, b *big.Int) bool {
+	return a.Cmp(b) > 0
+}
```

Comment:
These utils should also be moved somewhere generic?

Reply:
c699c67